gh-102978: Fix mock.patch function signatures for class and staticmethod decorators

tomasr8 opened the pull request:
Closes #102978 

As reported in the original [issue](https://github.com/python/cpython/issues/102978), `unittest.mock.patch` does not correctly extract function signatures from methods decorated with `@classmethod` and `@staticmethod` when `patch` is called with `autospec=True`.

This fix adds a special case to the extraction logic which uses the original decorated function via `__func__` to get the correct signature.

I also updated the relevant tests.


<!-- gh-issue-number: gh-102978 -->
* Issue: gh-102978
<!-- /gh-issue-number -->

sobolevn commented on Lib/unittest/mock.py:
```diff
@@ -98,6 +98,12 @@ def _get_signature_object(func, as_instance, eat_self):
         func = func.__init__
         # Skip the `self` argument in __init__
         eat_self = True
+    elif isinstance(func, classmethod | staticmethod):
```
```suggestion
    elif isinstance(func, (classmethod,  staticmethod)):
```

We should stick to tuples in this case :)

sobolevn commented on Misc/NEWS.d/next/Library/2023-04-03-23-44-34.gh-issue-102978.gy9eVk.rst:
```diff
@@ -0,0 +1,3 @@
+Fixes unittest.mock.patch not enforcing function signatures for methods
+decorated with @classmethod or @staticmethod when patch is called with
+autospec=True.
```
```suggestion
Fixes :func:`unittest.mock.patch` not enforcing function signatures for methods
decorated with ``@classmethod`` or ``@staticmethod`` when patch is called with
``autospec=True``.
```

sobolevn commented on Lib/test/test_unittest/testmock/testpatch.py:
```diff
@@ -989,12 +989,18 @@ def test_autospec_staticmethod(self):
             method.assert_called_once_with()
 
+            # Ensure that the original method's signature is checked correctly
+            self.assertRaises(TypeError, Foo.class_method, 42)
```
I think that this test is not enough. Please, create a new test with class and static methods that have some parameters. And check that these parameters do work :)

tomasr8 replied:
Thanks for reviewing this PR! I added more tests :)

sobolevn reviewed (commented):
Thanks!

tomasr8 committed:
Fix mock.patch function signatures for class and staticmethod decorators

Fixes unittest.mock.patch not enforcing function signatures for methods
decorated with @classmethod or @staticmethod when patch is called with
autospec=True.

sobolevn commented on Lib/test/test_unittest/testmock/testhelpers.py:
```diff
@@ -952,6 +952,30 @@ def __getattr__(self, attribute):
+            @staticmethod
+            def static_method(a, b=10, *, c): pass
+
+        Foo.static_method(1, 2, c=3)
```
We don't really need this line

sobolevn commented on Lib/test/test_unittest/testmock/testhelpers.py:
```diff
@@ -952,6 +952,30 @@ def __getattr__(self, attribute):
+            @classmethod
+            def class_method(cls, a, b=10, *, c): pass
+
+        Foo.class_method(1, 2, c=3)
```
And this

sobolevn reviewed (commented):
Almost done, in my opinion :)

sobolevn reviewed (commented):
LGTM, but you will need an approve from the code owner :)

cjw296 commented on Lib/test/test_unittest/testmock/testhelpers.py:
```diff
@@ -952,6 +952,26 @@ def __getattr__(self, attribute):
 
 
+    def test_autospec_signature_staticmethod(self):
+        # Regression test for gh-102978
```
It's a small thing, but these comments don't add much and will bitrot.
```suggestion
```

cjw296 commented on Lib/test/test_unittest/testmock/testhelpers.py:
```diff
@@ -952,6 +952,26 @@ def __getattr__(self, attribute):
+
+
+    def test_autospec_signature_classmethod(self):
+        # Regression test for gh-102978
```
```suggestion
```

cjw296 commented on Lib/test/test_unittest/testmock/testpatch.py:
```diff
@@ -996,6 +996,38 @@ def test_autospec_classmethod(self):
 
+    def test_autospec_staticmethod_signature(self):
+        # Regression test for gh-102978
+        # Patched methods which are decorated with @staticmethod should have the same signature
```
```suggestion
        # Patched methods which are decorated with @staticmethod should have the same signature
```

cjw296 commented on Lib/test/test_unittest/testmock/testpatch.py:
```diff
@@ -996,6 +996,38 @@ def test_autospec_classmethod(self):
+
+    def test_autospec_classmethod_signature(self):
+        # Regression test for gh-102978
+        # Patched methods which are decorated with @classmethod should have the same signature
```
```suggestion
        # Patched methods which are decorated with @classmethod should have the same signature
```

cjw296 reviewed (approved):
Really nice PR, thanks!

cjw296 committed:
Update Lib/test/test_unittest/testmock/testhelpers.py

cjw296 committed:
Update Lib/test/test_unittest/testmock/testhelpers.py

cjw296 committed:
Update Lib/test/test_unittest/testmock/testpatch.py

cjw296 committed:
Update Lib/test/test_unittest/testmock/testpatch.py

cjw296 committed:
Merge branch 'main' into gh-102978
